File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int lmax[] = new int[height.length];
        int rmax[] = new int[height.length];
        lmax[0] = height[0];
        for(int i = 1;i<lmax.length;i++)
            lmax[i] = Math.max(height[i],lmax[i-1]);
        rmax[rmax.length-1] = height[rmax.length-1];
        for(int i = rmax.length-2;i>=0;i--)
            rmax[i] = Math.max(height[i],rmax[i+1]);
        int res = 0;
        for(int i = 0;i<lmax.length;i++){
            res+= Math.min(rmax[i],lmax[i]) - height[i];
         //   System.out.print(res+" ");
        }
        return res;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& nums) {
        int n=nums.size();
        vector<int> f(n),b(n);
        
        f[0] = -1;
        for(int i=1;i<n;i++)
            f[i] = max(nums[i-1],f[i-1]);
        b[n-1]=-1;
        for(int i=n-2;i>=0;i--)
            b[i] = max(nums[i+1],b[i+1]);
        
        int ans=0;
        for(int i=0;i<n;i++){
            if(i == 0 || i == n-1)
                continue;
            if(min(f[i],b[i]) > nums[i])
                ans+=min(f[i],b[i])-nums[i];
        }
        return ans;
    }
};